jgrapht-core/src/main/java/org/jgrapht/alg/transform/LineGraphConverter.java

```diff
@@ -0,0 +1,132 @@
+ * corresponds to the arc set of $G$ and having an arc directed from an edge $e_1$ to an
+ * edge $e_2$ if in $G$, the head of $e_1$ meets the tail of $e_2$
+ *
+ * <p>
+ * More formally, let $G = (V, E)$ be a graph then its line graph $L(G)$ is such that
+ * <ul>
+ * <li>Each vertex of $L(G)$ represents an edge of $G$</li>
+ * <li>Two vertices of $L(G)$ are adjacent if and only if their corresponding edges share
+ * a common endpoint ("are incident") in $G$ </li>
+ * </ul>
+ * <p>
+ *
+ * @author Joris Kinable
+ * @author Nikhil Sharma
+ *
+ *
+ * @param <V> vertex type of input graph
+ * @param <E> edge type of input graph
+ * @param <EE> edge type of target graph
+ *
+ */
+public class LineGraphConverter<V, E, EE>
+{
+
+    private final Graph<V, E> graph;
+
+    /**
+     * Line Graph Converter
+     *
+     * @param graph graph to be converted
+     */
+    public LineGraphConverter(Graph<V, E> graph)
+    {
+        this.graph = Objects.requireNonNull(graph, "Graph cannot be null");
+    }
+
+    /**
+     * Constructs a line graph $L(G)$ of the input graph $G(V,E)$. If the input graph is directed, the result is a line digraph.
+     * The result is stored in the target graph.
+     *
+     * @param target target graph
+     */
+    public void convertToLineGraph(final Graph<E, EE> target)
+    {
+        this.convertToLineGraph(target, null);
+    }
+
+    /**
+     * Constructs a line graph of the input graph. If the input graph is directed, the result is a line digraph.
+     * The result is stored in the target graph. A weight function is provided to set edge weights of the line graph
+     * edges. Notice that the target graph must be a weighted graph for this to work. Recall that in a line graph
+     * $L(G)$ of a graph $G(V,E)$ there exists an edge $e$ between $e1\in E$ and $e2\in E$ if the head of $e1$ is incendent
+     * to the tail of $e2$. To determine the weight of $e$ in $L(G)$, the weight function takes as input $e1$ and $e2$.
+     *
+     * @param target target graph
+     * @param weightFunction weight function
+     */
+    public void convertToLineGraph(final Graph<E, EE> target, final BiFunction<E, E, Double> weightFunction)
+    {
+        Graphs.addAllVertices(target, graph.edgeSet());
+        if (graph.getType().isDirected()) {
+            for(V vertex : graph.vertexSet()) {
+                for (E e1 : graph.incomingEdgesOf(vertex)) {
+                    for (E e2 : graph.outgoingEdgesOf(vertex)) {
+                        EE edge = target.addEdge(e1, e2);
+                        if (weightFunction != null)
+                            target.setEdgeWeight(edge, weightFunction.apply(e1, e2));
+                    }
+                }
+            }
+        } else{ // undirected graph
+            for(V v : graph.vertexSet()) {
+                for (E e1 : graph.edgesOf(v)) {
+                    for (E e2 : graph.edgesOf(v)) {
+                        if (e1 != e2) {
```

Comment:
What about self-loops in the original graph?  It seems like they should result in self-loops in the line graph as well, although I don't see any mention of this one way or another in wikipedia or mathworld.

Reply:
I've added a note on self-loops. Self-loops are a bit 'weird' in the context of line graphs. I've kept the implementation consistent with https://networkx.github.io/documentation/networkx-1.10/reference/generated/networkx.generators.line.line_graph.html (see the note at the end).
In practice I think the user has to determine by himself what kind of behavior is desirable when creating a linegraph of a pseudograph.

Reply:
Got it.  The networkx rationale seems inconsistent between undirected and directed, but at least it's a pre-existing standard to follow, so I'm happy with that.